### PR TITLE
dmUpdate: add diurnal reduction

### PR DESCRIPTION
**Pull request by rem1776**
**Description**
Adds the diurnal reduction and test.

Fixes # (issue)

**How Has This Been Tested?**
Please describe the tests that you ran to verify your changes. Please also note
any relevant details for your test configuration (e.g. compiler, OS).  Include
enough information so someone can reproduce your tests.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: diag_manager/fms_diag_object.F90 [hunk 1]
```diff
@@ -203,9 +203,11 @@ integer function fms_register_diag_field_obj &
  class (fmsDiagFile_type), pointer :: fileptr !< Pointer to the diag_file
  class (fmsDiagField_type), pointer :: fieldptr !< Pointer to the diag_field
  class (fmsDiagOutputBuffer_type), pointer :: bufferptr !< Pointer to the output buffer
+ class (diagYamlFilesVar_type), pointer :: yamlfptr
  integer, allocatable :: file_ids(:) !< The file IDs for this variable
  integer :: i !< For do loops
  integer, allocatable :: diag_field_indices(:) !< indices where the field was found in the yaml
+ logical :: is_diurnal
```
Comment (uramirez8707) [inline diag_manager/fms_diag_object.F90:210]: It looks like this variable is not actually used/needed?

File: test_fms/diag_manager/test_time_diurnal.sh [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+printf "&test_diag_diurnal_nml \n test_case=0 \n mask_case=2 \n / \n" >> input.nml
+
+test_expect_success "monthly diurnal output with real mask" '
+  mpirun -n 6 ../test_diag_diurnal
+'
+test_expect_success "checking results for diurnal test with real mask" '
+  mpirun -n 6 ../check_time_diurnal
+'
+
+printf "&test_diag_diurnal_nml \n test_case=1 \n / \n" >> input.nml
```
Comment (uramirez8707) [inline test_fms/diag_manager/test_time_diurnal.sh:105]: I think you need to set `OMP_NUM_THREADS` to 2 here

**Review by uramirez8707 — APPROVED** [review]

File: diag_manager/include/fms_diag_reduction_methods.inc [hunk 3]
```diff
@@ -239,6 +242,7 @@ subroutine DO_TIME_SUM_UPDATE_(data_out, weight_sum, data_in, mask, is_masked, b
   integer :: i, j, k, l !< For looping
   real(FMS_TRM_KIND_) :: weight_loc, pow_loc !< local copies of optional arguments
   integer, parameter  :: kindl = FMS_TRM_KIND_ !< real kind size as set by macro
+  integer :: diurnal
```
Comment (thomas-robinson) [inline diag_manager/include/fms_diag_reduction_methods.inc:245]: needs documentation

**Review by thomas-robinson — CHANGES_REQUESTED** [review]
Please make sure all of the variables have descriptive documentation

**Review by thomas-robinson — APPROVED** [review]

**Pull request merged**